External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,507 @@
+#pragma once
+
+#include <atomic>
+#include <FEXCore/Utils/Allocator.h>
+#include <FEXCore/Utils/LogManager.h>
+
+#include <chrono>
+#include <cstddef>
+#include <list>
+#include <mutex>
+#include <sys/mman.h>
+
+namespace FEXCore::Utils {
+  /**
+   * @brief An intrusive thread pool allocator
+   *
+   * Requires coordination between the allocator and its clients to efficiently share memory allocations between threads.
+   *
+   * The `Client` in this case referring to the location in code allocating a `MemoryBuffer` from the allocator.
+   *   - The client must `Claim` a buffer to allocate it
+   *   - In claiming a buffer, the allocator is passed a `BufferOwnedFlag` that is updated by both the allocator and client.
+   *   - When the client is done with the buffer it must `Disown` or `Unclaim` the buffer.
+   *     - `Disown` the buffer when it is expected to be used again soon.
+   *       - This is relatively cheap.
+   *     - `Unclaim` when the buffer won't be used again for an extended period.
+   *       - This is expensive and requires a mutex shared between threads
+   *     - `FixedSizeDelayedObjectReclaimer` helper class provided to help with this.
+   *
+   * Once the client has disowned a buffer then the allocator is free to reclaim the buffer when another thread is trying to `Claim` a new buffer.
+   * The buffer getting claimed from a disowned client must have had its last use greater than the defined `DURATION` before it has a chance to get
+   * reclaimed by the Allocator.
+   *
+   * During buffer reclaiming is also when unclaimed buffers get freed. This means active threads are able to clean up idle thread's unused memory.
+   */
+  class IntrusiveThreadPoolAllocator {
```

Comment:
How about we rename this to  `IntrusivePooledAllocator` to make it clear it doesn't allocate from a `ThreadPool`?

Reply:
Resolved.